### PR TITLE
Update pubsub client with latest surface changes.

### PR DESCRIPTION
**Pull request by shinfan**
Pubsub now is compatible with GAX 0.0.9
More details about the surface change: 
https://github.com/googleapis/gax-java/commit/e27c1e5d4d64273ff291c781e3ec60862b9ebb8f

https://github.com/googleapis/gax-java/commit/5509053ba1425405348244e7ea86fedd150851ef

### CODE REVIEW

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 1]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
+    private SimpleCallSettings.Builder<Subscription, Subscription> createSubscriptionSettings;
+    private SimpleCallSettings.Builder<GetSubscriptionRequest, Subscription>
+        getSubscriptionSettings;
+    private PageStreamingCallSettings.Builder<
+            ListSubscriptionsRequest, ListSubscriptionsResponse, Subscription>
+        listSubscriptionsSettings;
+    private SimpleCallSettings.Builder<DeleteSubscriptionRequest, Empty> deleteSubscriptionSettings;
+    private SimpleCallSettings.Builder<ModifyAckDeadlineRequest, Empty> modifyAckDeadlineSettings;
+    private SimpleCallSettings.Builder<AcknowledgeRequest, Empty> acknowledgeSettings;
+    private SimpleCallSettings.Builder<PullRequest, PullResponse> pullSettings;
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'modifyPushConfigSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l218)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 2]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
+
+    private SimpleCallSettings.Builder<Subscription, Subscription> createSubscriptionSettings;
+    private SimpleCallSettings.Builder<GetSubscriptionRequest, Subscription>
+        getSubscriptionSettings;
+    private PageStreamingCallSettings.Builder<
+            ListSubscriptionsRequest, ListSubscriptionsResponse, Subscription>
+        listSubscriptionsSettings;
+    private SimpleCallSettings.Builder<DeleteSubscriptionRequest, Empty> deleteSubscriptionSettings;
+    private SimpleCallSettings.Builder<ModifyAckDeadlineRequest, Empty> modifyAckDeadlineSettings;
+    private SimpleCallSettings.Builder<AcknowledgeRequest, Empty> acknowledgeSettings;
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'pullSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l217)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 3]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
           };
+
+  public static class Builder extends ServiceApiSettings.Builder {
+    private final ImmutableList<ApiCallSettings.Builder> methodSettingsBuilders;
+
+    private SimpleCallSettings.Builder<Subscription, Subscription> createSubscriptionSettings;
+    private SimpleCallSettings.Builder<GetSubscriptionRequest, Subscription>
+        getSubscriptionSettings;
+    private PageStreamingCallSettings.Builder<
+            ListSubscriptionsRequest, ListSubscriptionsResponse, Subscription>
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'listSubscriptionsSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l213)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 4]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
+        listSubscriptionsSettings;
+    private SimpleCallSettings.Builder<DeleteSubscriptionRequest, Empty> deleteSubscriptionSettings;
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'modifyAckDeadlineSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l215)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 5]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
               return payload.getSubscriptionsList();
             }
           };
+
+  public static class Builder extends ServiceApiSettings.Builder {
+    private final ImmutableList<ApiCallSettings.Builder> methodSettingsBuilders;
+
+    private SimpleCallSettings.Builder<Subscription, Subscription> createSubscriptionSettings;
+    private SimpleCallSettings.Builder<GetSubscriptionRequest, Subscription>
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'getSubscriptionSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l210)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 6]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
               return payload.getSubscriptionsList();
             }
           };
+
+  public static class Builder extends ServiceApiSettings.Builder {
+    private final ImmutableList<ApiCallSettings.Builder> methodSettingsBuilders;
+
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'createSubscriptionSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l208)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 7]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
+    private SimpleCallSettings.Builder<Subscription, Subscription> createSubscriptionSettings;
+    private SimpleCallSettings.Builder<GetSubscriptionRequest, Subscription>
+        getSubscriptionSettings;
+    private PageStreamingCallSettings.Builder<
+            ListSubscriptionsRequest, ListSubscriptionsResponse, Subscription>
+        listSubscriptionsSettings;
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'deleteSubscriptionSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l214)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java [hunk 8]
```diff
@@ -336,4 +201,212 @@ public Object extractNextToken(ListSubscriptionsResponse payload) {
...
+    private SimpleCallSettings.Builder<DeleteSubscriptionRequest, Empty> deleteSubscriptionSettings;
+    private SimpleCallSettings.Builder<ModifyAckDeadlineRequest, Empty> modifyAckDeadlineSettings;
```
Comment (mziccard) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/SubscriberSettings.java]: Codacy Issue found: [Private field 'acknowledgeSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705098/issues/source?bid=3123109&fileBranchId=3225976#l216)

File: gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 9]
```diff
@@ -412,4 +285,196 @@ public long countBytes(PublishRequest request) {
...
         }
       };
+
+  public static class Builder extends ServiceApiSettings.Builder {
+    private final ImmutableList<ApiCallSettings.Builder> methodSettingsBuilders;
+
+    private SimpleCallSettings.Builder<Topic, Topic> createTopicSettings;
+    private BundlingCallSettings.Builder<PublishRequest, PublishResponse> publishSettings;
+    private SimpleCallSettings.Builder<GetTopicRequest, Topic> getTopicSettings;
+    private PageStreamingCallSettings.Builder<ListTopicsRequest, ListTopicsResponse, Topic>
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: Codacy Issue found: [Private field 'listTopicsSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705100/issues/source?bid=3123109&fileBranchId=3225976#l296)

File: gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 10]
```diff
@@ -412,4 +285,196 @@ public long countBytes(PublishRequest request) {
           return request.getSerializedSize();
         }
       };
+
+  public static class Builder extends ServiceApiSettings.Builder {
+    private final ImmutableList<ApiCallSettings.Builder> methodSettingsBuilders;
+
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: Codacy Issue found: [Private field 'createTopicSettings' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2381705100/issues/source?bid=3123109&fileBranchId=3225976#l292)

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherApi.java [hunk 11]
```diff
@@ -321,6 +309,7 @@ public PublishResponse publish(PublishRequest request) {
    * <!-- end manual edit -->
    *
    * @param topic The name of the topic to get.
```
Comment (aozarov) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherApi.java]: Nit, code generation comment: param description should start with lower case (like the description for throws, return,...) and does not need a period). For more see [this](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#tag).

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherApi.java [hunk 12]
```diff
@@ -409,6 +406,9 @@ private Topic getTopic(GetTopicRequest request) {
    *
    * <!-- manual edit -->
    * <!-- end manual edit -->
+   *
+   * @param topic The name of the topic that subscriptions are attached to.
+   * @throws ApiException if the remote call fails
    */
   public final Iterable<String> listTopicSubscriptions(String topic) {
```
Comment (aozarov) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherApi.java]: Any updates regarding paging?

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 13]
```diff
@@ -33,19 +33,17 @@
...
+import com.google.api.gax.core.RetrySettings;
 import com.google.api.gax.grpc.ApiCallSettings;
-import com.google.api.gax.grpc.ApiCallable;
-import com.google.api.gax.grpc.ApiCallable.Builder;
-import com.google.api.gax.grpc.ApiCallable.BundlableBuilder;
-import com.google.api.gax.grpc.ApiCallable.PageStreamingBuilder;
+import com.google.api.gax.grpc.BundlingCallSettings;
 import com.google.api.gax.grpc.BundlingDescriptor;
 import com.google.api.gax.grpc.BundlingSettings;
+import com.google.api.gax.grpc.PageStreamingCallSettings;
```
Comment (aozarov) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: I see you still call this one Streaming.... any issue or pending PR to fix it?

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 14]
```diff
@@ -412,4 +285,196 @@ public long countBytes(PublishRequest request) {
...
+    private BundlingCallSettings.Builder<PublishRequest, PublishResponse> publishSettings;
+    private SimpleCallSettings.Builder<GetTopicRequest, Topic> getTopicSettings;
+    private PageStreamingCallSettings.Builder<ListTopicsRequest, ListTopicsResponse, Topic>
+        listTopicsSettings;
+    private PageStreamingCallSettings.Builder<
+            ListTopicSubscriptionsRequest, ListTopicSubscriptionsResponse, String>
+        listTopicSubscriptionsSettings;
+    private SimpleCallSettings.Builder<DeleteTopicRequest, Empty> deleteTopicSettings;
+
+    private static final ImmutableMap<String, ImmutableSet<Status.Code>> RETRYABLE_CODE_DEFINITIONS;
```
Comment (aozarov) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: You may want to be aware of #883.
Summary: we think it makes sense to retry when:
idempotent AND error code suggests it makes sense to retry
OR
error code suggests that it makes sense to retry and that request was rejected

An example of the latter case would be a short burst quota exceeded.

File: gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 15]
```diff
@@ -412,4 +285,196 @@ public long countBytes(PublishRequest request) {
...
+      RETRYABLE_CODE_DEFINITIONS = definitions.build();
+    }
+
+    private static final ImmutableMap<String, RetrySettings.Builder> RETRY_PARAM_DEFINITIONS;
+
+    static {
+      ImmutableMap.Builder<String, RetrySettings.Builder> definitions = ImmutableMap.builder();
+      RetrySettings.Builder settingsBuilder = null;
+      settingsBuilder =
+          RetrySettings.newBuilder()
```
Comment (aozarov) [inline gcloud-java-pubsub/baseline/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: @eschapira Any reason Pub/Sub [SLA](https://cloud.google.com/pubsub/sla) does not specify "Back-off Requirements" like many other [services](https://cloud.google.com/datastore/sla).
Do these look correct to you?

File: gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java [hunk 16]
```diff
@@ -412,4 +285,196 @@ public long countBytes(PublishRequest request) {
...
+      RETRYABLE_CODE_DEFINITIONS = definitions.build();
+    }
+
+    private static final ImmutableMap<String, RetrySettings.Builder> RETRY_PARAM_DEFINITIONS;
+
+    static {
+      ImmutableMap.Builder<String, RetrySettings.Builder> definitions = ImmutableMap.builder();
+      RetrySettings.Builder settingsBuilder = null;
+      settingsBuilder =
+          RetrySettings.newBuilder()
```
Comment (aozarov) [inline gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/v1/PublisherSettings.java]: This seems to be an unnecessary line wrap.

**Comment by aozarov** [comment]
Looks much better. Some future comments/suggestions. Feel free to merge.

**Pull request merged**